src/smtsolvers/TheoryInterpolator.h

```diff
@@ -3,7 +3,139 @@
 
 #include "Global.h"
 #include "PtStore.h"
-#include <PartitionManager.h>
+#include "PartitionManager.h"
+#include "OsmtInternalException.h"
+
+#include <memory>
+#include <algorithm>
+
+class TermColorInfo {
+public:
+    virtual icolor_t getColorFor(PTRef term) = 0;
+    virtual ~TermColorInfo() = default;
+
+};
+
+class GlobalTermColorInfo : public TermColorInfo {
+public:
+    GlobalTermColorInfo(PartitionManager & pmanager, ipartitions_t mask) : pmanager(pmanager), mask(std::move(mask)) {}
+
+    icolor_t getColorFor(PTRef term) override {
+        auto const & termMask = pmanager.getIPartitions(term);
+        auto res = getColorForMask(termMask);
+        if (res == I_UNDEF) {
+            throw OsmtInternalException("No color detected for term");
+        }
+        return res;
+    }
+
+private:
+    PartitionManager & pmanager;
+    ipartitions_t mask;
+
+    icolor_t getColorForMask(ipartitions_t const & otherMask) {
+        bool isInA = (otherMask & mask) != 0;
+        bool isInB = (otherMask & ~mask) != 0;
+        if (isInA and not isInB) { return I_A; }
+        if (isInB and not isInA) { return I_B; }
+        if (isInA and isInB) { return I_AB; }
+        return I_UNDEF;
+    }
+};
+
+/*
+ * Stores color information for a set of terms given the colors of top-level term.
+ *
+ * Terms can be A-local, B-local, or AB-shared.
+ * Note: If a term f(x) is local, but both the function symbol and all the arguments are AB-shared,
+ *       then f(x) will also be stored as shared.
+ */
+class LocalTermColorInfo : public TermColorInfo {
+public:
+
+    template<typename TMap>
+    LocalTermColorInfo(TMap const & topLevelMap, Logic const & logic) {
+        termColors[logic.getTerm_true()] = I_AB;
+        termColors[logic.getTerm_false()] = I_AB;
+        computeColorsForAllSubterms(topLevelMap, logic);
+    }
+
+    icolor_t getColorFor(PTRef term) override {
+        auto it = termColors.find(term);
+        if (it == termColors.end()) {
+            throw OsmtInternalException("No color detected for term");
+        }
+        return it->second;
+    }
+
+private:
+    std::unordered_map<PTRef, icolor_t, PTRefHash> termColors;
+
+    template<typename TMap>
+    void computeColorsForAllSubterms(TMap const & topLevelColors, Logic const & logic) {
+        // MB: NOTE! If P(a) is A-local, but both symbols P and a are shared, than P(a) should be shared and not A-local
+        using entry_t = std::pair<const PTRef, icolor_t>;
+        auto colorUnion = [](icolor_t f, icolor_t s) { return static_cast<opensmt::icolor_t>(f | s); };
+        std::vector<entry_t> queue;
+        for (auto entry : topLevelColors) {
+            queue.push_back(entry);
+        }
+        std::unordered_map<SymRef, icolor_t, SymRefHash> symbolColors;
+        while (not queue.empty()) {
+            auto const & entry = queue.back();
```

Comment:
Maybe the reference is not necessary here?

Reply:
You would prefer taking it by value? Or something like
```
PTRef term = queue.back().first;
icolor_t colorToAssign = queue.back().second;
```
?

Reply:
Taken by value.